src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VCProjectSystemServices.cs

```diff
@@ -0,0 +1,407 @@
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+using MicrosoftBuildEvaluationProject = Microsoft.Build.Evaluation.Project;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Contains the information specific to a Visual Basic or C# project.
+    /// </summary>
+    internal class VCProjectSystemServices
+        : GlobalProjectServiceProvider
+        , INuGetProjectServices
+        , IProjectSystemCapabilities
+        , IProjectSystemReferencesReader
+        , IProjectSystemReferencesService
+    {
+        private static readonly Array _referenceMetadata;
+
+        private readonly IVsProjectAdapter _vsProjectAdapter;
+        private readonly IVsProjectThreadingService _threadingService;
+        private readonly Lazy<VCProject> _asVCProject;
+
+        private VCProject AsVCProject => _asVCProject.Value;
+
+        public bool SupportsPackageReferences => true;
+
+        #region INuGetProjectServices
+
+        public IProjectBuildProperties BuildProperties => _vsProjectAdapter.BuildProperties;
+
+        public IProjectSystemCapabilities Capabilities => this;
+
+        public IProjectSystemReferencesReader ReferencesReader => this;
+
+        public IProjectSystemReferencesService References => this;
+
+        public IProjectSystemService ProjectSystem => throw new NotSupportedException();
+
+        public IProjectScriptHostService ScriptService { get; }
+
+        #endregion INuGetProjectServices
+
+        static VCProjectSystemServices()
+        {
+            _referenceMetadata = Array.CreateInstance(typeof(string), 5);
+            _referenceMetadata.SetValue(ProjectItemProperties.IncludeAssets, 0);
+            _referenceMetadata.SetValue(ProjectItemProperties.ExcludeAssets, 1);
+            _referenceMetadata.SetValue(ProjectItemProperties.PrivateAssets, 2);
+            _referenceMetadata.SetValue(ProjectItemProperties.NoWarn, 3);
+            _referenceMetadata.SetValue(ProjectItemProperties.GeneratePathProperty, 4);
+        }
+
+        public VCProjectSystemServices(
+            IVsProjectAdapter vsProjectAdapter,
+            IComponentModel componentModel)
+            : base(componentModel)
+        {
+            Assumes.Present(vsProjectAdapter);
+
+            _vsProjectAdapter = vsProjectAdapter;
+
+            _threadingService = GetGlobalService<IVsProjectThreadingService>();
+            Assumes.Present(_threadingService);
+
+            _asVCProject = new Lazy<VCProject>(() => vsProjectAdapter.Project.Object as VCProject);
+
+            ScriptService = new VsProjectScriptHostService(vsProjectAdapter, this);
+        }
+
+        public async Task<IEnumerable<LibraryDependency>> GetPackageReferencesAsync(
+            NuGetFramework targetFramework, CancellationToken _)
+        {
+            Assumes.Present(targetFramework);
+
+            //await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
why couldn't _threadingService be used?

Reply:
I referenced it and it used NuGetUIThreadHelper.JoinableTaskFactory directly.

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/CpsProjectSystem.cs#L46

Reply:
I found out that they are actually the same object.